pkg/build/model.go

```diff
@@ -0,0 +1,419 @@
+func (buildInfo *BuildInfo) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var rawString string
+	err := unmarshal(&rawString)
+	if err == nil {
+		buildInfo.Name = rawString
+		return nil
+	}
+
+	var rawBuildInfo buildInfoRaw
+	err = unmarshal(&rawBuildInfo)
+	if err != nil {
+		return err
+	}
+
+	buildInfo.Name = rawBuildInfo.Name
+	buildInfo.Context = rawBuildInfo.Context
+	buildInfo.Dockerfile = rawBuildInfo.Dockerfile
+	buildInfo.Target = rawBuildInfo.Target
+	buildInfo.Args = rawBuildInfo.Args
+	buildInfo.Image = rawBuildInfo.Image
+	buildInfo.CacheFrom = rawBuildInfo.CacheFrom
+	buildInfo.ExportCache = rawBuildInfo.ExportCache
+	buildInfo.DependsOn = rawBuildInfo.DependsOn
+	buildInfo.Secrets = rawBuildInfo.Secrets
+	return nil
+}
+
+// BuildInfoRaw represents the build info for serialization
+type buildInfoRaw struct {
+	Secrets          BuildSecrets      `yaml:"secrets,omitempty"`
+	Name             string            `yaml:"name,omitempty"`
+	Context          string            `yaml:"context,omitempty"`
+	Dockerfile       string            `yaml:"dockerfile,omitempty"`
+	Target           string            `yaml:"target,omitempty"`
+	Image            string            `yaml:"image,omitempty"`
+	CacheFrom        cache.CacheFrom   `yaml:"cache_from,omitempty"`
+	Args             BuildArgs         `yaml:"args,omitempty"`
+	VolumesToInclude []VolumeMounts    `yaml:"-"`
+	ExportCache      cache.ExportCache `yaml:"export_cache,omitempty"`
+	DependsOn        BuildDependsOn    `yaml:"depends_on,omitempty"`
+}
+
+// MarshalYAML Implements the marshaler interface of the yaml pkg.
+func (buildInfo *BuildInfo) MarshalYAML() (interface{}, error) {
+	if buildInfo.Context != "" && buildInfo.Context != "." {
+		return buildInfoRaw(*buildInfo), nil
+	}
+	if buildInfo.Dockerfile != "" && buildInfo.Dockerfile != "./Dockerfile" {
+		return buildInfoRaw(*buildInfo), nil
+	}
+	if buildInfo.Target != "" {
+		return buildInfoRaw(*buildInfo), nil
+	}
+	if buildInfo.Args != nil && len(buildInfo.Args) != 0 {
+		return buildInfoRaw(*buildInfo), nil
+	}
+	return buildInfo.Name, nil
+}
+
+// UnmarshalYAML Implements the Unmarshaler interface of the yaml pkg.
+func (ba *BuildArgs) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	buildArgs := make(BuildArgs, 0)
+	result, err := getBuildArgs(unmarshal)
+	if err != nil {
+		return err
+	}
+	for key, value := range result {
+		buildArgs = append(buildArgs, BuildArg{Name: key, Value: value})
+	}
+	sort.SliceStable(buildArgs, func(i, j int) bool {
+		return strings.Compare(buildArgs[i].Name, buildArgs[j].Name) < 0
+	})
+	*ba = buildArgs
+	return nil
+}
+
+// MarshalYAML Implements the marshaler interface of the yaml pkg.
+func (v VolumeMounts) MarshalYAML() (interface{}, error) {
+	return v.RemotePath, nil
+}
+
+// UnmarshalYAML Implements the Unmarshaler interface of the yaml pkg.
+func (v *VolumeMounts) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
shall we have `VolumeMounts` and its marshalling in a dedicated `volume.go` to make this file smaller?